apisix/plugins/ldap-auth-advanced.lua

```diff
@@ -0,0 +1,308 @@
+
+local auth_header_lrucache = core.lrucache.new({
+    ttl = 300, count = 512
+})
+
+local schema = {
+    type = "object",
+    properties = {
+        ldap_uri = { type = "string" },
+        use_ldaps = {
+            type = "boolean",
+            default = false
+        },
+        use_starttls = {
+            type = "boolean",
+            default = false
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true,
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            maximum = 60000,
+            default = 3000,
+            description = "timeout in milliseconds",
+        },
+        keepalive = {type = "boolean", default = true},
+        keepalive_timeout = {type = "integer", minimum = 1000, default = 60000},
+        keepalive_pool_size = {type = "integer", minimum = 1, default = 5},
+        keepalive_pool_name = {type = "string", default = nil},
+        ldap_debug = {
+            type = "boolean",
+            default = false,
+        },
+        hide_credentials = {
+            type = "boolean",
+            default = false,
+        },
+        consumer_required = {
+            type = "boolean",
+            default = true,
+        },
+        user_dn_template = {
+            type = "string",
+        },
+        user_membership_attribute = {
+            type = "string",
+            default = "memberOf",
+        }
+    },
+    required = {"ldap_uri", "user_dn_template"}
+}
+
+
+local consumer_schema = {
+    type = "object",
+    properties = {
+        user_dn = { type = "string" },
+        group_dn = { type = "string" },
+    },
+    oneOf = {
+        {required = {"user_dn"}},
+        {required = {"group_dn"}},
+    }
+}
+
+local plugin_name = "ldap-auth-advanced"
+
+local _M = {
+    version = 0.1,
+    priority = 2541,
+    type = "auth",
+    name = plugin_name,
+    schema = schema,
+    consumer_schema = consumer_schema,
+}
+
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_CONSUMER then
+        return core.schema.check(consumer_schema, conf)
+    end
+
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return ok, err
+    end
+
+    -- ensure %s in template
+    if not str_find(conf.user_dn_template, "%s") or not ok then
+        return false, "User DN template doesn't contain " ..
+                        "the %s placeholder for the uid variable"
+    end
+
+    -- ensure only one %s in template
+    if not pcall(str_format, conf.user_dn_template, "username") then
+        return false, "User DN template has more than one " ..
+                        "placeholder %s for the username variable."
+    end
+
+    if conf.use_starttls and conf.use_ldaps then
+        return false, "STARTTLS and LDAPS cannot be open at the same time"
+    end
+
+    return true
+end
+
+
+local function extract_auth_header(authorization)
```

Comment:
This function is used multiple times in different plugins, can we put them in a unified location. Maybe record a todo